src/core/lib/compression/message_compress.h

```diff
@@ -24,13 +24,18 @@
 #include <grpc/impl/codegen/compression_types.h>
 #include <grpc/slice.h>
 
+struct gzip_compression_options {
```

Comment:
This doesn't really address my prior comment.  The problem here is that the `grpc_msg_compress()` API is intended to be algorithm-agnostic, so it does not make sense to pass into it anything that is specific to any one algorithm.  Moving the gzip-specific option into a gzip-specific struct doesn't really solve that problem, because in both cases, we're passing something specific to gzip into a function that is not specific to gzip.

I think we need an abstraction here to allow us to pass in the options in an opaque way, so that the parameters are not gzip-specific.  I suggest something like the following:

- In message_compress.h:
  ```
  namespace grpc_core {
  class CompressionOptions {
   public:
    virtual ~CompressionOptions() = default;
  };
  std::unique_ptr<CompressionOptions> MakeCompressionOptions(const grpc_channel_args* args);
  }  // namespace grpc_core
  
  int grpc_msg_compress(grpc_compression_algorithm algorithm,
                        grpc_slice_buffer* input, grpc_slice_buffer* output,
                        const grpc_core::CompressionOptions* options);
  ```
- In message_compress.cc:
  ```
  namespace grpc_core {
  class CompressionOptionsImpl : public CompressionOptions {
   public:
    explicit CompressionOptionsImpl(const grpc_channel_args* args)
        : gzip_compression_level_(
              DefaultGzipCompressionLevelFromChannelArgs(args)) {}

    int gzip_compression_level() const { return gzip_compression_level_; }
    
   private:
    int gzip_compression_level_;
  };
  
  std::unique_ptr<CompressionOptions> MakeCompressionOptions(const grpc_channel_args* args) {
    return absl::make_unique<CompressionOptionsImpl>(args);
  }
  }  // namespace grpc_core
  
  int grpc_msg_compress(grpc_compression_algorithm algorithm,
                        grpc_slice_buffer* input, grpc_slice_buffer* output,
                        const grpc_core::CompressionOptions* options) {
    if (!compress_inner(algorithm, input, output,
                        static_cast<grpc_core::CompressionOptionsImpl*>(options))) {
      copy(input, output);
      return 0;
    }
    return 1;
  }  
  ```
- In the message compress filter, in `ChannelData`, call `grpc_core::MakeCompressionOptions()` and save the result in a data member.  Then, in `CallData`, use that data member from `ChannelData` to pass the option to `grpc_msg_compress()`.

Reply:
maybe struct is more sample

Reply:
Yes, a struct is simpler, but getting the opaque API that we want here with a struct would make the casting less type-safe.  I think using a class interface is preferable here.